### ISSUE TITLE
Input lost on unhandled exception thrown in durable entities

### ISSUE THREAD

**Issue by asos-sarahwood** [Azure/azure-functions-durable-extension#1751]
### Description
When we throw an unhandled exception from within a durable entity we see a failure but lose the input. There are no input message found on the queues in the storage account.

We are using the function based syntax

### Expected behavior
As we have not overridden the default value for RollbackOnException we expected the state not to be persisted and the input to be retried.

### Actual behavior
We lose the input message

**Comment by sebastianburckhardt** [comment]
From your description, it appears that the implementation is working as designed: there is no automatic retry for entity operations if the application code throws an exception.

If retries are important in your situation, your best option at the moment is to implement your own retries.  You can use standard exception handling to catch the exception. To retry it, you can use a loop, or you can have the entity send a signal to itself (or perhaps a delayed signal).

**Comment by olitomlinson** [comment]
@sebastianburckhardt 

I wonder if the docs need to be more explicit about there not being any retry by default on an Entity Operation? I wrongly assumed there was a retry...

Also, given that entity operations are dequeued in batches, does that mean they are also ACKd in batches too?

So, given a batch of 50 entity operations (which may span N entities in the partition) if the 25th throws an exception, does that rollback all the state for entities 1-25?

**Comment by sebastianburckhardt** [comment]
>  wonder if the docs need to be more explicit about there not being any retry by default on an Entity Operation? 

I agree that more documentation on exception handling is needed. Note that the design is consistent with how we handle exceptions in activities. The runtime considers an operation that throws an exception as completed (just as for activities). The exception is considered the "result" of that operation and is propagated back to the caller (if the operation was a call and not a signal). If it is a signal, the exception cannot be handled by the signaller, but it is still visible in the traces.

This distinction between runtime-internal faults, and exceptions thrown by application code is very important for the programming model in general. Runtime-internal faults are retried transparently. But exceptions thrown by the application are not. The rationale is that the runtime does not really know the meaning of exceptions thrown from application code, e.g. it cannot know whether the exception reflects a transient fault and should be retried, or whether it is an issue that cannot be resolved by retrying and that should therefore be handled by an exception handler in the calling orchestration. 

> So, given a batch of 50 entity operations (which may span N entities in the partition) if the 25th throws an exception, does that rollback all the state for entities 1-25?

Depends on the setting of RollbackOnException. 

- RollbackOnException=true  (default)
   Only operation 25 is rolled back. All other operations (1-24, 26-50) are processed normally. 
   The rollback resets the entity state to what it was before starting operation 25, and suppresses any signals that were sent or orchestrations that were started by operation 25.

- RollbackOnException=false
   There is no rollback:  everything the operation 25 does before throwing the exception remains in effect (all entity state updates, signals sent, and orchestrations started). And operations 26-50 are also processed as usual.

One thing I realize as I am writing this is that `RollbackOnException` is probably not supported for out-of-proc languages like JavaScript or Python. @ConnorMcMahon or @davidmrdavid can probably clarify this. We may need separate docs for those, I have not had a close look at how they implement this.

**Comment by olitomlinson** [comment]
@sebastianburckhardt awesome thanks!

Follow up...

> RollbackOnException=true (default)
Only operation 25 is rolled back. All other operations (1-24, 26-50) are processed normally.
The rollback resets the entity state to what it was before starting operation 25, and suppresses any signals that were sent or orchestrations that were started by operation 25.

Are signals which may be destined for entities in a _different_ partition _also_ rolled back? [No caveats/nuances etc]

**Comment by davidmrdavid** [comment]
Thanks for the ping @sebastianburckhardt. I have personally not validated the `RollbackOnException` scenario in OOProc yet. I'm tracking that item in these 2 tickets:
- https://github.com/Azure/azure-functions-durable-python/issues/274
- https://github.com/Azure/azure-functions-durable-js/issues/250

**Comment by sebastianburckhardt** [comment]
> Are signals which may be destined for entities in a different partition also rolled back? [No caveats/nuances etc]

The rollback includes all the effects that the operation performs via the `IDurableEntityContext`. The partition does not matter.

Any effects performed in other ways (e.g. direct I/O calls, or using an injected IDurableClient) are not visible to the runtime and are not rolled back.

**Comment by YunchuWang** [comment]
Retested on 2026-09-09 against real Azure-hosted resources. **The reported behavior still reproduces, and matches the application-exception semantics explained above.**

Environment: Azure Functions v4 on Windows Consumption in West US 2, Azure Storage backend (not an emulator), .NET 8 in-process, function-based entity syntax. The deployed app reported `Microsoft.Azure.WebJobs.Extensions.DurableTask` **3.15.0**, the latest stable release for this programming model, and Functions host `4.852.100.26380`. No rollback override was supplied; the effective `RollbackEntityOperationsOnExceptions` value was `true`.

The failing operation reads its input, changes state, and then throws an unhandled `InvalidOperationException`, equivalent to:

```csharp
[FunctionName("Counter")]
public static void Counter([EntityTrigger] IDurableEntityContext context)
{
    switch (context.OperationName)
    {
        case "add":
            context.SetState(context.GetState<int>() + context.GetInput<int>());
            break;
        case "fail":
            context.SetState(context.GetState<int>() + context.GetInput<int>());
            throw new InvalidOperationException("intentional application failure");
    }
}
```

Observed results:

| Scenario | Result |
| --- | --- |
| Signal an existing entity with state 10, add 99 and throw (3 repetitions) | State rolls back to 10. A subsequent successful +1 signal produces 11. |
| Signal a new entity, create state 99 and throw (3 repetitions) | No entity state persists. A subsequent successful +1 signal produces 1. |
| Call the failing entity operation from an orchestrator | State rolls back to 10; the unhandled original exception reaches the orchestrator, which becomes `Failed`. |
| Send 50 ordered signals from an orchestrator, each adding 1, with operation 25 throwing | The other 49 updates persist; final state is 49. |

For execution counting, the repro wrote each operation's synthetic input and tentative state to a separate private blob before throwing. This external audit write intentionally survives entity rollback. All eight failed operations executed exactly once. During a 120-second observation window, there were no additional attempts and all four control queues plus the work-item queue reported zero messages. A confirmed app stop/start also produced no additional attempts; all entity states remained correct.

So the input is delivered to the entity, but an application exception completes the operation as a failure: **rollback does not mean requeue or automatic retry**. The lack of a queued input afterward is reproducible, not evidence that state rollback failed. Applications that need retries must implement them explicitly, as described in the maintainer response.

Current documentation: [operation rollback](https://learn.microsoft.com/azure/durable-task/durable-functions/durable-functions-dotnet-entities?pivots=in-proc#operation-rollback) and [entity error handling](https://learn.microsoft.com/azure/durable-task/common/durable-task-error-handling?pivots=durable-functions#handle-errors-in-entity-functions).

**Comment by YunchuWang** [comment]
Follow-up on 2026-09-10: **this behavior is not exclusive to .NET in-process.** I repeated the reproduction using .NET isolated on real Azure Functions and Azure Storage.

The deployed worker reported `FUNCTIONS_WORKER_RUNTIME=dotnet-isolated` and `Microsoft.Azure.Functions.Worker.Extensions.DurableTask` **1.19.0**, the latest stable worker extension. Environment: .NET 8.0.30, Windows Consumption in West US 2, Functions host `4.1054.250.26428`. The worker package resolved host-side Durable extension 3.15.0 and Durable Task .NET SDK 1.24.1. Entity operations used the function-based `TaskEntityDispatcher.DispatchAsync(operation => ...)` API with no rollback or retry overrides.

| Scenario | .NET isolated result |
| --- | --- |
| Existing entity: state 10, add 99 and throw; repeated 3 times | Failed change rolls back. A subsequent successful +1 signal produces 11. |
| New entity: create state 99 and throw; repeated 3 times | No state persists. A subsequent successful +1 signal produces 1. |
| Orchestrator calls the failing operation | Orchestrator becomes `Failed`; its output contains `EntityOperationFailedException` and the original application-error message. Entity state rolls back to 10. |
| 50 ordered signals, with operation 25 throwing | The other 49 operations succeed; final state is 49. |

External private-blob audit records confirmed receipt of each synthetic input. Across the complete run, all eight failed operations executed exactly once. There were 67 audited non-read operations in total. No additional attempts appeared during a 120-second observation window, and all four control queues plus the work-item queue reported zero messages. After a confirmed app stop/start, the worker process changed and the same states and execution counts remained.

The result matches the earlier in-process reproduction: **entity state rollback does not requeue the failed input or automatically retry an application exception.** The notable model difference here is exception wrapping (`EntityOperationFailedException` in isolated), not automatic retry behavior.

Leaving this open rather than closing it as an in-process-only issue: that condition is not met. As the original maintainer explanation notes, the observed no-automatic-retry behavior is by design.
